File: lib/services/scrapers/defacto_scraper.dart
import 'dart:convert';
import 'package:html/dom.dart' as dom;
import 'base_scraper.dart';

class DefactoScraper extends BaseProductScraper {
  @override
  String get domain => 'defacto.com.tr';

  @override
  bool canHandle(String url) {
    return url.toLowerCase().contains('defacto.com.tr');
  }

  @override
  String? scrape({
    required dom.Document document,
    required String url,
    required bool Function(String urlString) isLogoUrl,
    required String? Function(String? imageUrl, String pageUrl) resolveImageUrl,
    required void Function(String message) log,
  }) {
    // 1. og:image tag'i dene (Defacto için son derece güvenlidir)
    final ogImage = document.querySelector('meta[property="og:image"]')?.attributes['content'];
    if (ogImage != null && ogImage.isNotEmpty) {
      final resolved = resolveImageUrl(ogImage, url);
      if (resolved != null && !isLogoUrl(resolved)) {
        log('✅ Defacto görseli og:image ile bulundu: $resolved');
        return resolved;
      }
    }

    // 2. DOM Seçicileri (Fallback)
    final imgSelectors = [
      '.product-card__image img',
      '.product-image img',
      'img[class*="product"]',
      'main img',
    ];
    for (final selector in imgSelectors) {
      final element = document.querySelector(selector);
      if (element != null) {
        final src = element.attributes['src'] ?? element.attributes['data-src'];
        if (src != null && src.isNotEmpty && !isLogoUrl(src)) {
          final resolved = resolveImageUrl(src, url);
          if (resolved != null) {
            log('✅ Defacto görseli DOM ile bulundu: $resolved');
            return resolved;
          }
        }
      }
    }

    return null;
  }

  @override
  String? scrapeTitle(dom.Document document) {
    // 1. Script bloğundan çekmeyi dene (Öncelikli)
    final scripts = document.querySelectorAll('script');
    for (final script in scripts) {
      final text = script.text;
      if (text.contains('PRODUCT_DETAIL_LASTVISITED') || text.contains('PRODUCT_DETAIL_INFO')) {
        final match = RegExp(r'"?ProductVariantMiniProductName"?\s*:\s*"([^"]+)"').firstMatch(text) ??
                      RegExp(r'"?Name"?\s*:\s*"([^"]+)"').firstMatch(text) ??
                      RegExp(r'"?name"?\s*:\s*"([^"]+)"').firstMatch(text);
        if (match != null) {
          final title = match.group(1);
          if (title != null && title.isNotEmpty) {
            return _decodeUnicode(title);
          }
        }
      }
    }

    // 2. DOM Seçicileri (Fallback)
    final titleEl = document.querySelector('h1.product-card__title') ?? 
                    document.querySelector('.product-title') ?? 
                    document.querySelector('h1');
    if (titleEl != null) {
      return titleEl.text.trim();
    }
    return null;
  }

  @override
  Future<double?> scrapePrice(dom.Document document) async {
    // 1. Script bloğundan (Öncelikli)
    final scripts = document.querySelectorAll('script');
    for (final script in scripts) {
      final text = script.text;
      if (text.contains('PRODUCT_DETAIL_LASTVISITED') || text.contains('CampaignBadge')) {
        // Fiyat belirleme algoritması:
        // A) Sepette indirimli fiyat (DiscountPrice) varsa bunu kullan
        final discountPriceMatch = RegExp(r'"?DiscountPrice"?\s*:\s*([0-9.]+)').firstMatch(text);
        if (discountPriceMatch != null) {
          final val = double.tryParse(discountPriceMatch.group(1)!);
          if (val != null && val > 0) return val;
        }

        // B) DataLayer altındaki CampaignDiscountedPrice varsa bunu kullan
        final campaignDiscountMatch = RegExp(r'"?CampaignDiscountedPrice"?\s*:\s*([0-9.]+)').firstMatch(text);
        if (campaignDiscountMatch != null) {
          final val = double.tryParse(campaignDiscountMatch.group(1)!);
          if (val != null && val > 0) return val;
        }

        // C) Ürünün standart indirimli fiyatı (ProductVariantMiniDiscountedPriceInclTax)
        final miniDiscountMatch = RegExp(r'"?ProductVariantMiniDiscountedPriceInclTax"?\s*:\s*"([0-9.]+)"').firstMatch(text);
        if (miniDiscountMatch != null) {
          final val = double.tryParse(miniDiscountMatch.group(1)!);
          if (val != null && val > 0) return val;
        }
      }
    }

    // 2. DOM Seçicileri (Fallback)
    final priceSelectors = [
      '.product-price__discount',
      '.product-price',
      'span[class*="price"]',
    ];
    for (final selector in priceSelectors) {
      final priceEl = document.querySelector(selector);
      if (priceEl != null) {
        final parsed = parsePriceText(priceEl.text);
        if (parsed != null && parsed > 0) return parsed;
      }
    }

    return null;
  }

  @override
  String? scrapeDescription(dom.Document document) {
    // 1. DOM meta description
    final descEl = document.querySelector('meta[name="description"]') ?? 
                   document.querySelector('meta[property="og:description"]');
    if (descEl != null) {
      final content = descEl.attributes['content']?.trim();
      if (content != null && content.isNotEmpty && content.toLowerCase() != 'null') {
        return content;
      }
    }
    return null;
  }

  String _decodeUnicode(String input) {
    var text = input;
    text = text.replaceAllMapped(RegExp(r'\\u([0-9a-fA-F]{4})'), (match) {
      final hexCode = match.group(1)!;
      final charCode = int.parse(hexCode, radix: 16);
      return String.fromCharCode(charCode);
    });
    text = text
      .replaceAll('&#x131;', 'ı')
      .replaceAll('&#x130;', 'İ')
      .replaceAll('&#x15F;', 'ş')
      .replaceAll('&#x15E;', 'Ş')
      .replaceAll('&#xE7;', 'ç')
      .replaceAll('&#xC7;', 'Ç')
      .replaceAll('&#xF6;', 'ö')
      .replaceAll('&#xD6;', 'Ö')
      .replaceAll('&#xFC;', 'ü')
      .replaceAll('&#xDC;', 'Ü')
      .replaceAll('&#x11F;', 'ğ')
      .replaceAll('&#x11E;', 'Ğ')
      .replaceAll('&quot;', '"')
      .replaceAll('&amp;', '&');
    return text;
  }

  @override
  List<String> scrapeBreadcrumbs(dom.Document document) {
    final scripts = document.querySelectorAll('script');
    final productTitle = scrapeTitle(document) ?? '';

    for (final script in scripts) {
      final type = script.attributes['type']?.trim().toLowerCase();
      if (type == 'application/ld+json') {
        try {
          final sanitizedText = script.text.replaceAll('\r\n', ' ').replaceAll('\n', ' ').replaceAll('\r', ' ');
          final data = jsonDecode(sanitizedText);
          final breadcrumbs = _extractBreadcrumbsFromJson(data, productTitle);
          if (breadcrumbs.isNotEmpty) {
            return breadcrumbs;
          }
        } catch (_) {}
      }
    }

    // DOM Fallback
    final breadcrumbElements = document.querySelectorAll('.breadcrumb a, .breadcrumbs a, .defacto-breadcrumb a, .breadcrumb-list a');
    if (breadcrumbElements.isNotEmpty) {
      final List<String> list = [];
      for (final el in breadcrumbElements) {
        final text = el.text.trim();
        if (text.isNotEmpty) {
          final lower = text.toLowerCase();
          if (lower != 'anasayfa' && lower != 'ana sayfa' && !lower.contains('defacto') && text.toLowerCase() != productTitle.toLowerCase().trim() && text.length < 50) {
            list.add(text);
          }
        }
      }
      if (list.isNotEmpty) return list;
    }

    return [];
  }

  List<String> _extractBreadcrumbsFromJson(dynamic json, String productTitle) {
    if (json is Map) {
      // 1. BreadcrumbList kontrolü
      if (json['@type'] == 'BreadcrumbList' || json['@type'] == 'http://schema.org/BreadcrumbList') {
        final items = json['itemListElement'];
        if (items is List) {
          final List<String> breadcrumbs = [];
          for (final item in items) {
            if (item is Map) {
              String? name;
              if (item['name'] != null) {
                name = item['name'].toString().trim();
              } else if (item['item'] is Map && item['item']['name'] != null) {
                name = item['item']['name'].toString().trim();
              }

              if (name != null && name.isNotEmpty) {
                final lowerName = name.toLowerCase();
                if (lowerName != 'anasayfa' && lowerName != 'ana sayfa' && !lowerName.contains('defacto')) {
                  if (lowerName != productTitle.toLowerCase().trim() && name.length < 50) {
                    breadcrumbs.add(name);
                  }
                }
              }
            }
          }
          if (breadcrumbs.isNotEmpty) return breadcrumbs;
        }
      }

      // 2. Product category kontrolü
      if (json['@type'] == 'Product' || json['@type'] == 'http://schema.org/Product' ||
          json['@type'] == 'ProductGroup' || json['@type'] == 'http://schema.org/ProductGroup') {
        final categoryField = json['category'];
        if (categoryField is String && categoryField.isNotEmpty) {
          final parts = categoryField
              .split(RegExp(r'\s*>\s*|\s*/\s*'))
              .map((e) => e.trim())
              .where((e) => e.isNotEmpty)
              .where((e) {
                final lower = e.toLowerCase();
                return lower != 'anasayfa' && lower != 'ana sayfa' && !lower.contains('defacto') && lower != productTitle.toLowerCase().trim() && e.length < 50;
              })
              .toList();
          if (parts.isNotEmpty) return parts;
        } else if (categoryField is Map) {
          final nameVal = categoryField['name'];
          if (nameVal is String && nameVal.isNotEmpty) {
            final parts = nameVal
                .split(RegExp(r'\s*>\s*|\s*/\s*'))
                .map((e) => e.trim())
                .where((e) => e.isNotEmpty)
                .where((e) {
                  final lower = e.toLowerCase();
                  return lower != 'anasayfa' && lower != 'ana sayfa' && !lower.contains('defacto') && lower != productTitle.toLowerCase().trim() && e.length < 50;
                })
                .toList();
            if (parts.isNotEmpty) return parts;
          }
        }
      }

      // Recursive arama
      for (final value in json.values) {
        if (value is Map || value is List) {
          final res = _extractBreadcrumbsFromJson(value, productTitle);
          if (res.isNotEmpty) return res;
        }
      }
    } else if (json is List) {
      for (final item in json) {
        final res = _extractBreadcrumbsFromJson(item, productTitle);
        if (res.isNotEmpty) return res;
      }
    }
    return [];
  }
}

File: cloud-run-bot/scrapers/defacto_scraper.js
/**
 * DeFacto Scraper (Node.js port)
 */
const BaseProductScraper = require('./base_scraper');

class DefactoScraper extends BaseProductScraper {
  get domain() { return 'defacto.com.tr'; }

  canHandle(url) {
    return url.toLowerCase().includes('defacto.com.tr');
  }

  _decodeUnicode(str) {
    try {
      return str.replace(/\\u([\dA-Fa-f]{4})/g, (_, grp) => String.fromCharCode(parseInt(grp, 16)));
    } catch (_) {
      return str;
    }
  }

  scrapeImage($, url) {
    // 1. og:image
    const ogImg = $('meta[property="og:image"]').attr('content');
    if (ogImg && !this.isLogoUrl(ogImg)) {
      const r = this.resolveImageUrl(ogImg, url);
      if (r) return r;
    }
    // 2. DOM
    const imgSelectors = [
      '.product-card__image img',
      '.product-image img',
      'img[class*="product"]',
      'main img'
    ];
    for (const sel of imgSelectors) {
      const el = $(sel).first();
      const src = el.attr('src') || el.attr('data-src');
      if (src && !this.isLogoUrl(src)) {
        const r = this.resolveImageUrl(src, url);
        if (r) return r;
      }
    }
    return null;
  }

  scrapeTitle($) {
    // 1. Script
    const scripts = $('script');
    for (let i = 0; i < scripts.length; i++) {
      const text = $(scripts[i]).text() || '';
      if (text.includes('PRODUCT_DETAIL_LASTVISITED') || text.includes('PRODUCT_DETAIL_INFO')) {
        const match = text.match(/"?ProductVariantMiniProductName"?\s*:\s*"([^"]+)"/) ||
                      text.match(/"?Name"?\s*:\s*"([^"]+)"/) ||
                      text.match(/"?name"?\s*:\s*"([^"]+)"/);
        if (match) {
          return this.unescapeHtml(this._decodeUnicode(match[1]));
        }
      }
    }
    // 2. DOM
    const el = $('h1.product-card__title, .product-title, h1').first();
    if (el.length) return el.text().trim();
    return null;
  }

  scrapePrice($) {
    // 1. DOM campaing-base-price (Sepette / Kampanyalı indirimli fiyat)
    const campaingEl = $('[class*="campaing-base-price"], .campaing-base-price, .product-price__discount').first();
    if (campaingEl.length) {
      const val = this.parsePriceText(campaingEl.text());
      if (val && val > 0) return val;
    }

    // 2. Script bloğundan
    const scripts = $('script');
    for (let i = 0; i < scripts.length; i++) {
      const text = $(scripts[i]).text() || '';
      if (text.includes('PRODUCT_DETAIL_LASTVISITED') || text.includes('CampaignBadge')) {
        // A) DiscountPrice
        const m1 = text.match(/"?DiscountPrice"?\s*:\s*([0-9.]+)/);
        if (m1) {
          const val = parseFloat(m1[1]);
          if (!isNaN(val) && val > 0) return val;
        }
        // B) CampaignDiscountedPrice
        const m2 = text.match(/"?CampaignDiscountedPrice"?\s*:\s*([0-9.]+)/);
        if (m2) {
          const val = parseFloat(m2[1]);
          if (!isNaN(val) && val > 0) return val;
        }
        // C) ProductVariantMiniDiscountedPriceInclTax
        const m3 = text.match(/"?ProductVariantMiniDiscountedPriceInclTax"?\s*:\s*"([0-9.]+)"/);
        if (m3) {
          const val = parseFloat(m3[1]);
          if (!isNaN(val) && val > 0) return val;
        }
      }
    }

    // 3. JSON-LD
    const product = this.findProductJsonLd($);
    if (product) {
      const p = this.extractPriceFromProductJson(product);
      if (p && p > 0) return p;
    }

    // 4. DOM fallback
    const priceSelectors = [
      '.product-price',
      '.product-info__price--new',
      'span[class*="price"]'
    ];
    for (const sel of priceSelectors) {
      const el = $(sel).first();
      if (el.length) {
        const val = this.parsePriceText(el.text());
        if (val && val > 0) return val;
      }
    }
    return null;
  }

  scrapeOriginalPrice($, currentPrice) {
    if (!currentPrice || currentPrice <= 0) return null;

    // 1. DOM lined-base-price (İndirimsiz çizili fiyat)
    const linedEl = $('[class*="lined-base-price"], .lined-base-price').first();
    if (linedEl.length) {
      const txt = linedEl.text().trim();
      const val = this.parsePriceText(txt);
      if (val && val > currentPrice) return val;
    }

    // 2. JSON-LD price (DeFacto JSON-LD genellikle indirim öncesi liste fiyatını verir)
    const product = this.findProductJsonLd($);
    if (product && product.offers && product.offers.price != null) {
      const p = parseFloat(product.offers.price);
      if (!isNaN(p) && p > currentPrice) return p;
    }

    // 3. Fallback selectors
    let candidates = [];
    const selectors = [
      '[class*="lined-base-price"]',
      '.lined-base-price',
      'del',
      's',
      '.old-price',
      '.original-price'
    ];
    for (const selector of selectors) {
      $(selector).each((_, el) => {
        const txt = $(el).text().trim();
        if (txt.includes('TL') || txt.includes('₺')) {
          const parsed = this.parsePriceText(txt);
          if (parsed !== null && parsed > currentPrice && parsed <= currentPrice * 5) {
            candidates.push(parsed);
          }
        }
      });
    }

    if (candidates.length === 0) return null;

    candidates.sort((a, b) => b - a);
    return candidates[0];
  }

  scrapeDescription($) {
    const descEl = $('meta[name="description"], meta[property="og:description"]').first();
    return descEl.length ? descEl.attr('content')?.trim() : null;
  }

  scrapeBreadcrumbs($) {
    const title = this.scrapeTitle($) || '';
    const breadcrumbs = this.extractBreadcrumbsFromJsonLd($, title, 'defacto');
    if (breadcrumbs && breadcrumbs.length > 0) return breadcrumbs;

    // DOM Fallback
    const list = [];
    $('.breadcrumb a, .breadcrumbs a, .defacto-breadcrumb a, .breadcrumb-list a').each((_, el) => {
      const text = $(el).text().trim();
      if (text) {
        const lower = text.toLowerCase();
        if (lower !== 'anasayfa' && lower !== 'ana sayfa' && !lower.includes('defacto') && lower !== title.toLowerCase().trim() && text.length < 50) {
          list.push(text);
        }
      }
    });
    return list;
  }

  scrapeRating($) {
    // 1. JSON-LD
    const product = this.findProductJsonLd($);
    if (product) {
      const rating = this.extractRatingFromProductJson(product);
      if (rating && (rating.ratingValue != null || rating.ratingCount != null)) {
        return rating;
      }
    }

    // 2. DOM
    const ratingEl = $('[itemprop="ratingValue"], meta[property="product:rating:value"], .rating-score, .pdp-rating-value').first();
    let ratingValue = null;
    let ratingCount = null;

    if (ratingEl.length) {
      const txt = ratingEl.is('meta') ? ratingEl.attr('content') : ratingEl.text();
      if (txt) {
        const p = parseFloat(txt.trim().replace(',', '.'));
        if (!isNaN(p) && p > 0 && p <= 5.0) ratingValue = p;
      }
    }

    const countEl = $('[itemprop="reviewCount"], [itemprop="ratingCount"], .review-count, .rating-count').first();
    if (countEl.length) {
      const txt = countEl.is('meta') ? countEl.attr('content') : countEl.text();
      if (txt) {
        const m = /(\d+)/.exec(txt);
        if (m) {
          const c = parseInt(m[1]);
          if (!isNaN(c) && c > 0) ratingCount = c;
        }
      }
    }

    if (ratingValue || ratingCount) {
      return { ratingValue, ratingCount };
    }

    return { ratingValue: null, ratingCount: null };
  }

  scrapeBrand($) {
    const product = this.findProductJsonLd($);
    if (product) {
      const brand = this.extractBrandFromProductJson(product);
      if (brand) return brand;
    }
    const metaBrand = $('meta[property="product:brand"], meta[name="brand"], .product-brand, [data-brand]').first();
    if (metaBrand.length) {
      const txt = metaBrand.is('meta') ? metaBrand.attr('content') : (metaBrand.attr('data-brand') || metaBrand.text());
      if (txt && txt.trim()) return txt.trim();
    }
    return null;
  }
}

module.exports = DefactoScraper;
